python/activator/logger.py

```diff
@@ -23,6 +23,61 @@
 
 import json
 import logging
+import os
+
+import lsst.daf.butler as daf_butler
+
+try:
+    import lsst.log as lsst_log
+except ModuleNotFoundError:
+    lsst_log = None
+
+
+def _parse_log_levels(spec):
+    """Parse a string description of logging levels.
+
+    Parameters
+    ----------
+    spec : `str`
+        A string consisting of space-separated pairs of logger=level
+        specifications. No attempt is made to validate whether the string
+        conforms to this format. "." refers to the root logger.
+
+    Returns
+    -------
+    levels : `list` [`tuple` [`str`, `str`]]
+        A list of tuples whose first element is a logger (root logger
+        represented by `None`) and whose second element is a logging level.
+    """
+    levels = [tuple(s.split("=", maxsplit=1)) for s in spec.split(None)]
+    return [(k if k != "." else None, v) for k, v in levels]
+
+
+def _set_lsst_logging_levels():
+    """Set up standard logging levels for LSST code.
+
+    This function consistently sets both the Python logger and lsst.log.
+    """
+    # Don't call CliLog.initLog because we need different formatters from what
+    # Butler assumes.
+    default_levels = [(None, "WARNING"),  # De-prioritize output from 3rd-party packages.
+                      ("lsst", "INFO"),   # Capture output from Middleware and pipeline.
+                      ]
+
+    log_request = os.environ.get("SERVICE_LOG_LEVELS", "")
+    daf_butler.cli.cliLog.CliLog.setLogLevels(default_levels + _parse_log_levels(log_request))
+
+
+def _channel_all_to_pylog():
+    """Set up redirection of lsst.log and warning output to the Python logger.
+
+    This ensures that all service output is formatted consistently, making it easier to parse.
+    """
+    if lsst_log is not None:
+        lsst_log.configure_pylog_MDC("DEBUG", MDC_class=None)
```

Comment:
Is what happens that the root logger gets set to DEBUG level here, and later the  `(None, "WARNING")` in `_set_lsst_logging_levels()` (line 63) resets it to WARNING?  So the level here doesn't really matter?

Reply:
I'll admit I cargo-culted this from `daf_butler`, but it's possible to change the root logging level from the envvar, and I think we'd want the log4cxx-Python bridge to work regardless.

Reply:
You need to call this to get forwarding from C++ logs to python logging. I'm pretty sure that if you set this to INFO or WARNING then you will not be able to ever get DEBUG logs out of C++ (I might be wrong on this). The explicit setting of the log level later means that using DEBUG here is not a problem.